_posts/2021-02-18-survey/survey.Rmd

```diff
@@ -0,0 +1,314 @@
+
+##### "No Python"
+
+The number-one concern with deep learning from R, for survey respondents, clearly has to do not with R but with Python. This topic appeared in various forms, the most frequent being frustration over how hard it can be, dependent on the environment, to get Python dependencies for TensorFlow/Keras correct. (It also appeared as enthusiasm for `torch`, which of course we are very happy about.)
+
+Let me clarify and add some context.
+
+TensorFlow is a Python framework (nowadays subsuming Keras, which is why I'll be addressing both of those as "TensorFlow" for simplicity) that is made available from R through packages `tensorflow` and `keras` . As with other Python libraries, objects are imported and accessible via `reticulate` . While `tensorflow` provides the low-level access, `keras` brings idiomatic-feeling, nice-to-use wrappers that let you forget about the chain of dependencies involved.
+
+On the other hand, `torch`, a recent addition to *mlverse* software, is an R port of PyTorch that does not delegate to Python. Instead, its R layer directly calls into `libtorch`, the C++ library behind PyTorch. In that way, it is like a lot of high-duty R packages, making use of C++ for performance reasons.
+
+Now, this is not the place for recommendations. Here are a few thoughts though.
+
+Clearly, as one respondent remarked, as of today the `torch` ecosystem does not offer functionality on par with TensorFlow, and for that to change time and -- hopefully! more on that below -- your, the community's, help is needed. Why? Because `torch` is so young, for one; but also, there is a "systemic" reason! With TensorFlow, as we can access any symbol via the `tf` object, it is always possible, if inelegant, to do from R what you see done in Python. Respective R wrappers nonexistent, quite a few blog posts (see, e.g., <https://blogs.rstudio.com/ai/posts/2020-04-29-encrypted_keras_with_syft/>, or [A first look at federated learning with TensorFlow](https://blogs.rstudio.com/ai/posts/2020-04-08-tf-federated-intro/)) relied on this!
+
+Switching to the topic of `tensorflow`'s Python dependencies causing problems with installation, my experience (from GitHub issues, plus adding in my own) has been that difficulties are quite system-dependent. On some OSes, complications seem to appear more often than on others; and low-control (to the individual user) environments like HPC clusters can make things especially difficult. In any case though, I have to (unfortunately) admit that when installation problems appear, they can be very tricky to solve.
+
+On to the next point (and I promise, this was by far the longest subsection).
+
+##### `tidymodels` integration
+
+The second most-frequent mention clearly was the wish for tighter `tidymodels` integration. Here, we wholeheartedly agree. As of today, there is no automated way to accomplish this for `torch` models generically, but it can be done for specific model implementations.
```

Comment:
```suggestion
The second most frequent mention clearly was the wish for tighter `tidymodels` integration. Here, we wholeheartedly agree. As of today, there is no automated way to accomplish this for `torch` models generically, but it can be done for specific model implementations.
```

Reply:
Most is actually an adverb here, so you don't need a hyphen.